tests/test_workers.py

```diff
@@ -0,0 +1,238 @@
+uvicorn_workers = pytest.importorskip("uvicorn.workers", reason="requires gunicorn")
+
+
+class Process(subprocess.Popen):
+    client: httpx.Client
+    output: IO[bytes]
+
+    def read_output(self) -> str:
+        self.output.seek(0)
+        return self.output.read().decode()
+
+
+@pytest.fixture(
+    params=(
+        pytest.param(uvicorn_workers.UvicornWorker, marks=pytest.mark.subprocess),
+        pytest.param(uvicorn_workers.UvicornH11Worker, marks=pytest.mark.subprocess),
+    )
+)
+def worker_class(request: pytest.FixtureRequest) -> str:
+    """Gunicorn worker class names to test.
+
+    This is a parametrized fixture. When the fixture is used in a test, the test
+    will be automatically parametrized, running once for each fixture parameter. All
+    tests using the fixture will be automatically marked with `pytest.mark.subprocess`.
+
+    https://docs.pytest.org/en/latest/how-to/fixtures.html
+    https://docs.pytest.org/en/latest/proposals/parametrize_with_fixtures.html
+    """
+    worker_class = request.param
+    return f"{worker_class.__module__}.{worker_class.__name__}"
+
+
+@pytest.fixture(
+    params=(
+        pytest.param(False, id="TLS off"),
+        pytest.param(True, id="TLS on"),
+    )
+)
+def gunicorn_process(
+    request: pytest.FixtureRequest,
+    tls_ca_certificate_pem_path: str,
+    tls_ca_ssl_context: SSLContext,
+    tls_certificate_private_key_path: str,
+    tls_certificate_server_cert_path: str,
+    unused_tcp_port: int,
+    worker_class: str,
+) -> Generator[Process, None, None]:
+    """Yield a subprocess running a Gunicorn arbiter with a Uvicorn worker.
+
+    An instance of `httpx.Client` is available on the `client` attribute.
+    Output is saved to a temporary file and accessed with `read_output()`.
+    """
+    app = "tests.test_main:app"
+    bind = f"127.0.0.1:{unused_tcp_port}"
+    use_tls: bool = request.param
+    args = [
+        "gunicorn",
+        "--bind",
+        bind,
+        "--graceful-timeout",
+        "1",
+        "--log-level",
+        "debug",
+        "--worker-class",
+        worker_class,
+        "--workers",
+        "1",
+    ]
+    if use_tls is True:
+        args_for_tls = [
+            "--ca-certs",
+            tls_ca_certificate_pem_path,
+            "--certfile",
+            tls_certificate_server_cert_path,
+            "--keyfile",
+            tls_certificate_private_key_path,
+        ]
+        args.extend(args_for_tls)
+        base_url = f"https://{bind}"
+        verify: SSLContext | bool = tls_ca_ssl_context
+    else:
+        base_url = f"http://{bind}"
+        verify = False
+    args.append(app)
+    client = httpx.Client(base_url=base_url, verify=verify)
+    output = tempfile.TemporaryFile()
+    with Process(args, stdout=output, stderr=output) as process:
+        time.sleep(1)
+        assert not process.poll()
+        process.client = client
+        process.output = output
+        yield process
```

Comment:
At first glance, it seems a little unusual to capture server output with a `TemporaryFile` like this, but it works well, especially for asserting on test output. I'm happy to consider alternatives that use [pytest's built-in capabilities](https://docs.pytest.org/en/latest/how-to/capture-stdout-stderr.html) like `capfd` or `capsys`, but I found the `TemporaryFile` approach to be more reliable.

See a similar example in Prefect's server code (PrefectHQ/prefect#7948), which uses Uvicorn :sparkles: :unicorn: :sparkles:.